lib/Optimizer/Transforms/LiftArrayAlloc.cpp

```diff
@@ -128,30 +151,44 @@ class AllocaPattern : public OpRewritePattern<cudaq::cc::AllocaOp> {
       for (auto &useuse : user->getUses()) {
         auto *useuser = useuse.getOwner();
         if (auto ist = dyn_cast<quake::InitializeStateOp>(useuser)) {
+          rewriter.setInsertionPointAfter(useuser);
           LLVM_DEBUG(llvm::dbgs() << "replaced init_state\n");
           assert(conGlobal && "global must be defined");
           rewriter.replaceOpWithNewOp<quake::InitializeStateOp>(
               ist, ist.getType(), ist.getTargets(), conGlobal);
           continue;
         }
         if (auto load = dyn_cast<cudaq::cc::LoadOp>(useuser)) {
+          rewriter.setInsertionPointAfter(useuser);
           LLVM_DEBUG(llvm::dbgs() << "replaced load\n");
           rewriter.replaceOpWithNewOp<cudaq::cc::ExtractValueOp>(
               load, eleTy, conArr,
               ArrayRef<cudaq::cc::ExtractValueArg>{offset});
           continue;
         }
-        if (isa<cudaq::cc::StoreOp>(useuser))
-          rewriter.eraseOp(useuser);
+        if (isa<cudaq::cc::StoreOp>(useuser)) {
+          toErase.push_back(useuser);
+        }
         isLive = true;
       }
       if (!isLive)
-        rewriter.eraseOp(user);
+        toErase.push_back(user);
     }
     if (toGlobal) {
+      rewriter.setInsertionPointAfter(alloc);
       rewriter.replaceOp(alloc, conGlobal);
     } else {
-      rewriter.eraseOp(alloc);
+      toErase.push_back(alloc);
+    }
+
+    for (auto *op : toErase) {
+      if (op->getUses().empty()) {
+        rewriter.eraseOp(op);
+      } else {
+        module.emitOpError("LiftArrayAlloc failed to remove quake.init_state "
```

Comment:
Can you provide a test case that reaches this error?

It appears the only ops on the list are cc::StoreOp which will never have users.

Reply:
I've run into it with loops that create arrays (like when python inlines the code that converts arrays to the simulation precision, or if a function that has a loop for initializing array is inlined)

Reply:
Removed the error